crates/ty_python_semantic/src/ast_node_ref.rs

```diff
@@ -52,34 +53,38 @@ impl<T> AstNodeRef<T> {
     /// Dereferencing the `node` can result in undefined behavior if `parsed` isn't the
     /// [`ParsedModuleRef`] to which `node` belongs. It's the caller's responsibility to ensure that
     /// the invariant `node belongs to parsed` is upheld.
-    pub(super) unsafe fn new(parsed: ParsedModuleRef, node: &T) -> Self {
+    pub(super) unsafe fn new(module_ref: &ParsedModuleRef, node: &T) -> Self {
         Self {
-            parsed,
-            node: std::ptr::NonNull::from(node),
+            module_ptr: module_ref.module().as_ptr(),
+            index: node.node_index().clone(),
+            _node: PhantomData,
         }
     }
 
     /// Returns a reference to the wrapped node.
     ///
-    /// Note that this method will panic if the provided module is from a different file or Salsa revision
-    /// than the module this node was created with.
-    pub fn node<'ast>(&self, parsed: &'ast ParsedModuleRef) -> &'ast T {
-        debug_assert!(Arc::ptr_eq(self.parsed.as_arc(), parsed.as_arc()));
+    /// Note that this method will panic if the provided module is from a different file or Salsa
+    /// revision than the module this node was created with.
+    pub fn node<'ast>(&self, module_ref: &'ast ParsedModuleRef) -> &'ast T {
+        debug_assert_eq!(module_ref.module().as_ptr(), self.module_ptr);
 
-        // SAFETY: Holding on to `parsed` ensures that the AST to which `node` belongs is still
-        // alive and not moved.
-        unsafe { self.node.as_ref() }
+        // Note that the module pointer is guaranteed to be stable within the Salsa
+        // revision, so the file contents cannot have changed by the above assertion.
```

Comment:
It's not clear to me why this is the case, given that we collect ASTs mid-revision.

Reply:
`ParsedModule` is an `Arc<ArcSwap<IndexedModule>>`. That outer `Arc` never changes, just like before. The garbage collection is invisible to Salsa.